### PR TITLE
[FEATURE]  Infrastructure for JavaScript modules and importmaps

### PR DESCRIPTION
**Pull request by linawolf**
resolves: https://github.com/TYPO3-Documentation/Changelog-To-Doc/issues/5
releases: main

### CODE REVIEW

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 1]
```diff
@@ -6,29 +6,111 @@
...
 ========================
 
-..  todo: Add information on ES6 and its registration
-    https://github.com/TYPO3-Documentation/Changelog-To-Doc/issues/5
-
-..  seealso::
-    This chapter is still work in progress. For now see the changelog:
+..  versionchanged:: 12.0
+    Starting with TYPO3 v12.0 JavaScript ES6 modules may be used instead of
+    AMD modules, both in backend and frontend context.
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:11]: Why mentioning frontend? I can use it also in v11 in frontend if I want. It sounds, that only v12 is capable of using ES6 in frontend - which is not true.

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 2]
```diff
@@ -6,29 +6,111 @@
...
+
+JavaScript node-js style path resolutions are managed by
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:13]:
```suggestion
JavaScript Node.js style path resolutions are managed by
```

"Node.js", see: https://nodejs.org/en/

But why mentioning Node.js? It is a standard which Node.js (and others) are implementing.

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 3]
```diff
@@ -6,29 +6,111 @@
...
+`importmaps <https://wicg.github.io/import-maps/>`_, which allow web
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:14]: The URL is a redirect to https://html.spec.whatwg.org/multipage/webappapis.html#import-maps

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 4]
```diff
@@ -6,29 +6,111 @@
...
+pages to control the behavior of JavaScript imports.
+
+By the time of writing importmaps are supported natively by Google Chrome,
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:17]:
```suggestion
By the time of writing, importmaps are supported natively by Google Chrome,
```

"By the time of writing": either add a date, like "November 2022" or add the browser versions you are referring to. The reader has no clue when it was written.

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 5]
```diff
@@ -6,29 +6,111 @@
...
+a polyfill is available for Firefox and Safari and included by TYPO3 core
+and applied whenever an importmap is emitted.
+
+For security reasons importmap configuration is only emitted when the modules
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:21]:
```suggestion
For security reasons, importmap configuration is only emitted when the modules
```

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 6]
```diff
@@ -6,29 +6,111 @@
...
+are actually used, that means when a module has been added to the current
+page response via via :php:`PageRenderer->loadJavaScriptModule()` or
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:23]:
```suggestion
page response via :php:`PageRenderer->loadJavaScriptModule()` or
```

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 7]
```diff
@@ -6,29 +6,111 @@
...
+:php:`JavaScriptRenderer->addJavaScriptModuleInstruction()`.
+Exposing all module configurations is possible via
+:php:`JavaScriptRenderer->includeAllImports()`, but that should only be
+done in backend context for logged in users, to avoid disclosing installed
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:27]:
```suggestion
done in backend context for logged-in users to avoid disclosing installed
```

File: Documentation/ApiOverview/Backend/JavaScript/ES6/_JavaScriptModulesExtended.php [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+            'path' => 'EXT:my_extension/Resources/Public/JavaScript/',
+            // Exclude files of the following folders from being import-mapped
+            'exclude' => [
+                'EXT:my_extension/Resources/Public/JavaScript/Contrib/',
+                'EXT:my_extension/Resources/Public/JavaScript/Overrides/',
+            ],
+        ],
+        // Adding a third party package
+        'thirdpartypkg' => 'EXT:my_extension/Resources/Public/JavaScript/Contrib/thidpartypkg/index.js',
+        'thidpartypkg/' => 'EXT:my_extension/Resources/Public/JavaScript/Contrib/thirdpartypkg/',
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/_JavaScriptModulesExtended.php:16]:
```suggestion
        'thirdpartypkg/' => 'EXT:my_extension/Resources/Public/JavaScript/Contrib/thirdpartypkg/',
```

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 9]
```diff
@@ -6,29 +6,111 @@
...
-..  note::
-    This chapter is still work in progress.
+A module can be added to the current page response either via
+:php:`PageRenderer` or as :php:`JavaScriptModuleInstruction` via
+:php:`JavaScriptRenderer`:
+
+..  literalinclude:: _PageRendererJavaScriptLoading.php
+    :caption: EXT:my_extension/Classes/SomeNamespace/SomeClass.php
+
+In Fluid template the `includeJavaScriptModules` property of the
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:57]:
```suggestion
In a Fluid template the `includeJavaScriptModules` property of the
```

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 10]
```diff
@@ -6,29 +6,111 @@
...
+:html:`<f:be.pageRenderer>` ViewHelper may be used:
+
+..  literalinclude:: _BackendFluidTemplate.html
+    :caption: EXT:my_extension/Resources/Private/Backend/Templates/SomeTemplate.html
+
+Some tips on ES6
+================
+
+In the TYPO3 Core there are no ES6 JavaScript files created directly. JavaScript
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:66]:
```suggestion
No ES6 JavaScript files are created directly in the TYPO3 Core. JavaScript
```

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 11]
```diff
@@ -6,29 +6,111 @@
...
+is created as TypeScript module which is then converted to ES6 JavaScript
+during the build process. However TypeScript and ES6 are quite similar, you
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:68]:
```suggestion
during the build process. However, TypeScript and ES6 are quite similar, you
```

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 12]
```diff
@@ -6,29 +6,111 @@
...
+can therefore look into those files for reference. The TypeScript files can be
+found on GitHub at
+`Build/Sources/TypeScript <https://github.com/TYPO3/typo3/tree/main/Build/Sources/TypeScript>`__.
+
+For an example on a ES6 JavaScript file have a look at the JavaScript example in
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:73]:
```suggestion
For examples of an ES6 JavaScript file have a look at the JavaScript example in
```

Too much "example" in this sentence, maybe "have a look at the JavaScript part in LinkHandler Tutorial or in the Notification API".

Why "LinkHandler Tutorial" and not "link handler tutorial"?

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 13]
```diff
@@ -6,29 +6,111 @@
...
+the :ref:`LinkHandler Tutorial <tutorial_backend_link_handler_javascript>` or
+the example in the :ref:`Notification API <notification_api>`.
+
+For a real life example on how to introduce ES6 modules into a large extension
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:77]:
```suggestion
For a practical example on how to introduce ES6 modules into a large extension
```

"real life" sounds that other examples are just theoretical.

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 14]
```diff
@@ -6,29 +6,111 @@
...
+see this commit for EXT:news: `[TASK] Add support for TYPO3 v12 ES6
+modules <https://github.com/bnf/news/commit/f8e196b67ceaa2f56699fbf464080dde668ad526>`__.
+
+Using JQuery
+------------
+
+There is a general tendency to remove dependencies to jQuery in the Core whenever
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:84]: Hmm, would rephrase that, something like: "In the TYPO3 Core usage of jQuery is eliminated step-by-step as the functionality is provided by native JavaScript nowadays."

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 15]
```diff
@@ -6,29 +6,111 @@
...
+    import $ from 'jquery';
 
 .. _requirejs-migration:
 
 Migration from RequireJS
 ========================
 
+..  note:: For general information on how to migrate from AMD (RequireJS) to
+    ES6 see this article: `Arthur Yidi: Migrate JavaScript Modules From AMD
+    to ES6 <https://arthuryidi.com/migrate-amd-modules/>`__. Here we talk about
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:101]: "Here we talk ..." -> "This is about the TYPO3-specific details."

We don't talk here, we read ;-)

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 16]
```diff
@@ -6,29 +6,111 @@
...
+    the TYPO3 specific details.
+
+RequireJS is shimmed to prefer ES6 modules if available, allowing any extension
+to ship ES6 modules by providing an importmap configuration in
+:file:`Configuration/JavaScriptModules.php` while providing full backwards
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:106]:
```suggestion
:file:`Configuration/JavaScriptModules.php` while providing full backward
```

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 17]
```diff
@@ -6,29 +6,111 @@
...
+compatibility support for extensions that load modules via RequireJS.
+
+Existing RequireJS modules can load new ES6 modules via a bridge that
+prefers ES6 modules over traditional RequireJS AMD modules. This allows
+extensions authors to migrate to ES6 without breaking dependencies that
+used to load a module of that extension via RequireJS.
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:112]:
```suggestion
previously loaded a module of that extension via RequireJS.
```

**Review by brotkrueml — CHANGES_REQUESTED** [review]

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst (hunk 1 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:11]: It is written like that in the changelog entry

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst (hunk 2 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:13]: It is written like that in the changelog entry, how would you call that

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst (hunk 3 above)
Comment (linawolf) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:14]:
```suggestion
`importmaps <https://html.spec.whatwg.org/multipage/webappapis.html#import-maps>`_, which allow web
```

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst (hunk 14 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:84]: Ok, go ahead

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 18]
```diff
@@ -6,29 +6,111 @@
...
+    import $ from 'jquery';
 
 .. _requirejs-migration:
 
 Migration from RequireJS
 ========================
 
+..  note:: For general information on how to migrate from AMD (RequireJS) to
+    ES6 see this article: `Arthur Yidi: Migrate JavaScript Modules From AMD
+    to ES6 <https://arthuryidi.com/migrate-amd-modules/>`__. Here we talk about
```
Comment (linawolf) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:101]:
```suggestion
    to ES6 <https://arthuryidi.com/migrate-amd-modules/>`__. This is about 
```

File: Documentation/ApiOverview/Backend/JavaScript/ES6/_JavaScriptModulesExtended.php (hunk 8 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/JavaScript/ES6/_JavaScriptModulesExtended.php:16]: Then the key would be the same like above. Also the example comes like that from the changelog, you are sure about this change?

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 19]
```diff
@@ -6,29 +6,111 @@
...
-    This chapter is still work in progress. For now see the changelog:
+..  versionchanged:: 12.0
+    Starting with TYPO3 v12.0 JavaScript ES6 modules may be used instead of
+    AMD modules, both in backend and frontend context.
+
+JavaScript node-js style path resolutions are managed by
+`importmaps <https://html.spec.whatwg.org/multipage/webappapis.html#import-maps>`_, which allow web
+pages to control the behavior of JavaScript imports.
+
+By the time of writing importmaps are supported natively by Google Chrome,
```
Comment (linawolf) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:17]:
```suggestion
In november 2022 importmaps are supported natively by Google Chrome,
```

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst (hunk 2 above)
Comment (brotkrueml) [reply Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:13]: I don't understand what that part means (but I am not a JS guy) ... then leave it.

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 20]
```diff
@@ -6,29 +6,111 @@
...
-    This chapter is still work in progress. For now see the changelog:
+..  versionchanged:: 12.0
+    Starting with TYPO3 v12.0 JavaScript ES6 modules may be used instead of
+    AMD modules, both in backend and frontend context.
+
+JavaScript node-js style path resolutions are managed by
+`importmaps <https://html.spec.whatwg.org/multipage/webappapis.html#import-maps>`_, which allow web
+pages to control the behavior of JavaScript imports.
+
+In november 2022 importmaps are supported natively by Google Chrome,
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:17]:
```suggestion
In November 2022 importmaps are supported natively by Google Chrome,
```

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 21]
```diff
@@ -6,29 +6,111 @@
...
+..  versionchanged:: 12.0
+    Starting with TYPO3 v12.0 JavaScript ES6 modules may be used instead of
+    AMD modules, both in backend and frontend context.
+
+JavaScript node-js style path resolutions are managed by
+`importmaps <https://html.spec.whatwg.org/multipage/webappapis.html#import-maps>`_, which allow web
+pages to control the behavior of JavaScript imports.
+
+In November 2022 importmaps are supported natively by Google Chrome,
+a polyfill is available for Firefox and Safari and included by TYPO3 core
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:18]:
```suggestion
a polyfill is available for Firefox and Safari and included by TYPO3 Core
```

**Review by brotkrueml — APPROVED** [review]
